maserol/predictAbundKa.py

```diff
@@ -1,32 +1,85 @@
+"""
+Currently runs polyfc to compare with SpaceX data with 6 receptors, 117 subjects, and 14 antigens.
+Polyfc is ran with initial guesses for abundance (117x1) and (14x1)  and Ka for each receptor (6x1).
+Polyfc output is compared to SpaceX data and cost function is minimzied through scipy.optimize.minimize.
+Total fitting parameters = 147
+"""
 import numpy as np
 from valentbind import polyfc
+from scipy.optimize import minimize
+import matplotlib.pyplot as plt
 
 
-def initial_AbundKa(flatCube, n_ab=1):
+def initial_AbundKa(cube, n_ab=1):
     """
     generate abundance and Ka matrices from random values
-    flatCube.shape == n_rec * (n_subj * n_Ag)
+    cube.shape == n_subj * n_rec * n_Ag
     """
-    R_guess = np.random.randint(4, high=11, size=(n_ab, flatCube.shape[1]), dtype=int)
-    Ka_guess = np.random.randint(1e8, high=9e8, size=(flatCube.shape[0], n_ab), dtype=int)
-    return R_guess, Ka_guess
+    R_subj_guess = np.random.lognormal(size=(cube.shape[0], n_ab))
+    R_Ag_guess = np.random.lognormal(size=(cube.shape[2], n_ab))
+    Ka_guess = np.random.lognormal(6, size=(cube.shape[1], n_ab))
+    return R_subj_guess, R_Ag_guess, np.log(Ka_guess)
 
 
-def infer_Lbound(R, Ka, L0=1e-9, KxStar=1e-12):
+def infer_Lbound(R_subj, R_Ag, Ka, L0=1e-9, KxStar=1e-12):
     """
     pass the matrices generated above into polyfc, run through each receptor
     and ant x sub pair and store in matrix same size as flatten
     """
-    Lbound_guess = np.zeros((Ka.shape[0], R.shape[1]))
+    Lbound_cube = np.zeros((R_subj.shape[0], Ka.shape[0], R_Ag.shape[0]))
+    # Lbound_guess = 6x1638
     LigC = np.array([1])
-    for ii in range(Lbound_guess.shape[0]):
-        for xx in range(Lbound_guess.shape[1]):
-            Lbound_guess[ii, xx] = polyfc(L0, KxStar, 2, R[:, xx], LigC, Ka[[ii], :])[0]
-    return Lbound_guess
+    Ka = np.exp(Ka[:, np.newaxis])
+    RR = np.einsum("ij,kj->ijk", R_subj, R_Ag)
```

Comment:
@erbozich this implements the low-rank receptor abundance assumption (equivalent to using the KR).